File: neetcode150/17_math_and_geometry/0054_spiral_matrix.py
"""
=====================================================================
LeetCode 54 : Spiral Matrix                                   (Medium)
https://leetcode.com/problems/spiral-matrix/
Category   : Math & Geometry
---------------------------------------------------------------------
PROBLEM
    Given an m×n matrix, return every element in SPIRAL order:
    right along the top edge, down the right edge, left along the
    bottom, up the left edge — shrinking inward until every cell
    has been visited exactly once.
---------------------------------------------------------------------
INTUITION
    A spiral is repeated BOUNDARY PEELING: walk one full loop around
    the outer rectangle (top → right → bottom → left), then shrink
    the rectangle by one cell on every side and repeat. When the top
    boundary passes the bottom (or left passes right) we are done.
---------------------------------------------------------------------
APPROACH 1 — Boundary shrinking with 4 pointers (BEST)
    Track top, bottom, left, right. Sweep each of the four edges of
    the current layer in turn, shrinking its bound after the sweep.
    Two guards (top <= bottom / left <= right) keep single-row and
    single-column layers from being double-visited.
    Time  Complexity : O(m·n)   Space Complexity : O(1) extra
    (the answer list of m·n values is the required output itself).
---------------------------------------------------------------------
APPROACH 2 — Layer-by-layer recursive peeling
    Same spiral as a recursion: peel the outer layer, recurse on the
    inner (m-2)×(n-2) rectangle. Same O(m·n) work; the recursion
    stack costs O(min(m, n)/2) memory instead of O(1).
    Time  Complexity : O(m·n)   Space Complexity : O(min(m, n)/2)
---------------------------------------------------------------------
REAL LIFE
    • Raster-to-vector scanners and image steganography read pixels
      in spiral order; printers (e.g. test pages) trace spirals.
    • Sparse matrix iterators and proximity searches (k-d tree
      neighbor rings) peel rings in this exact pattern.
=====================================================================
"""

from __future__ import annotations

import time
import tracemalloc

from typing import List


# =====================================================================
# APPROACH 1 : Boundary shrinking with 4 pointers
# =====================================================================
class SolutionOne:
    """
    Purpose : Return all matrix elements in spiral order.
    Inputs  : matrix — m×n list of lists of int.
    Output  : List[int] with elements in spiral order.
    """

    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        out: List[int] = []

        # The four shrinking boundaries of the current layer.
        top, bottom = 0, len(matrix) - 1
        left, right = 0, len(matrix[0]) - 1

        # Keep peeling layers until the rectangle collapses.
        while top <= bottom and left <= right:

            # 1) top row, left → right.
            for c in range(left, right + 1):
                out.append(matrix[top][c])
            top += 1                    # top edge consumed → shrink

            # 2) right column, top → bottom.
            for r in range(top, bottom + 1):
                out.append(matrix[r][right])
            right -= 1                  # right edge consumed → shrink

            # 3) bottom row, right → left. Guard: a single-row matrix
            #    would otherwise revisit cells swept in step 1.
            if top <= bottom:
                for c in range(right, left - 1, -1):
                    out.append(matrix[bottom][c])
                bottom -= 1

            # 4) left column, bottom → top. Guard: single-column case.
            if left <= right:
                for r in range(bottom, top - 1, -1):
                    out.append(matrix[r][left])
                left += 1

        return out


# =====================================================================
# APPROACH 2 : Layer-by-layer recursive peeling
# =====================================================================
class SolutionTwo:
    """
    Purpose : Return all matrix elements in spiral order.
    Inputs  : matrix — m×n list of lists of int.
    Output  : List[int] with elements in spiral order.
    """

    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        out: List[int] = []
        self._peel(matrix, 0, len(matrix) - 1, 0, len(matrix[0]) - 1, out)
        return out

    def _peel(self, matrix: List[List[int]], t: int, b: int,
              l: int, r: int, out: List[int]) -> None:
        """
        Purpose : Append one spiral layer to `out`, then recurse inward.
        Inputs  : matrix — the grid; t/b/l/r — current layer bounds;
                  out — accumulator filled in spiral order.
        Output  : None (out is appended in place).
        """
        if t > b or l > r:
            return                     # layer collapsed → done

        out.extend(matrix[t][l:r + 1])                     # top →
        out.extend(matrix[row][r] for row in range(t + 1, b + 1))  # right ↓
        if t < b:
            out.extend(matrix[b][c] for c in range(r - 1, l - 1, -1))  # bottom ←
        if l < r:
            out.extend(matrix[row][l] for row in range(b - 1, t, -1))  # left ↑

        # Recurse one layer deeper on every side.
        self._peel(matrix, t + 1, b - 1, l + 1, r - 1, out)


# =====================================================================
# BENCHMARK — both approaches on 3 matrices; time + peak memory (KB)
# =====================================================================
if __name__ == "__main__":
    tests = [
        [[1, 2, 3], [4, 5, 6], [7, 8, 9]],
        [[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12]],
        [[1], [2], [3], [4]],
    ]
    expected = [
        [1, 2, 3, 6, 9, 8, 7, 4, 5],
        [1, 2, 3, 4, 8, 12, 11, 10, 9, 5, 6, 7],
        [1, 2, 3, 4],
    ]

    def run(label: str, solve) -> bool:
        """
        Purpose : Run one solution class over every test and print a row.
        Inputs  : label — display name of the approach.
                  solve — callable(matrix) -> List[int].
        Output  : bool — True when every test passed.
        """
        tracemalloc.start()
        t0 = time.perf_counter()
        ok = True
        for i, (mat, want) in enumerate(zip(tests, expected)):
            got = solve(mat)
            pass_ = got == want
            ok = ok and pass_
            print(f"  test {i + 1} ({len(mat)}x{len(mat[0])}): {got}  {'PASS' if pass_ else 'FAIL'}")
        t1 = time.perf_counter()
        _, peak = tracemalloc.get_traced_memory()
        tracemalloc.stop()
        print(f"{label} : {'PASS' if ok else 'FAIL'}  time = {(t1 - t0) * 1e6:.1f} us  peak-mem = {peak // 1024} KB")
        return ok

    ok1 = run("Approach 1 (boundary shrinking)    ", lambda m: SolutionOne().spiralOrder(m))
    ok2 = run("Approach 2 (recursive layer peel)  ", lambda m: SolutionTwo().spiralOrder(m))

    print("PASS : both approaches produce the exact spiral order on all tests."
          if ok1 and ok2
          else "FAIL : at least one test failed.")

File: neetcode150/17_math_and_geometry/0054_spiral_matrix.cpp
/*
 * =====================================================================
 * LeetCode 54 : Spiral Matrix                                   (Medium)
 * https://leetcode.com/problems/spiral-matrix/
 * Category   : Math & Geometry
 * ---------------------------------------------------------------------
 * PROBLEM
 *   Given an m×n matrix, return every element in SPIRAL order:
 *   right along the top edge, down the right edge, left along the
 *   bottom, up the left edge — shrinking inward until all cells
 *   are visited.
 * ---------------------------------------------------------------------
 * INTUITION
 *   A spiral is just repeated boundary peeling: walk one full loop
 *   around the outer rectangle (top → right → bottom → left), then
 *   shrink the rectangle by one on every side and repeat. When the
 *   top boundary passes the bottom (or left passes right), we are
 *   done — every cell was claimed exactly once.
 * ---------------------------------------------------------------------
 * APPROACH 1 — Boundary shrinking with 4 pointers (BEST)
 *   Keep top, bottom, left, right. Loop: sweep top row left→right,
 *   then right column top→bottom, then bottom row right→left, then
 *   left column bottom→top; shrink each bound after its sweep.
 *   Time  Complexity : O(m·n)   Space Complexity : O(1) extra
 *   (the returned vector of m·n values is the required output, not
 *   auxiliary space).
 *   WHY it works: each boundary defines one straight segment of the
 *   spiral; shrinking them in order guarantees every cell is visited
 *   once and the corner cells are never double-visited (guards like
 *   top <= bottom check the degenerate single-row/single-column case).
 * ---------------------------------------------------------------------
 * APPROACH 2 — Layer-by-layer (recursive peeling)
 *   Same spiral, expressed as a recursion: peel the outer layer,
 *   then recurse on the inner (m-2)×(n-2) rectangle. Same O(m·n)
 *   work; the call stack costs O(min(m, n)/2) memory instead of O(1).
 *   Time  Complexity : O(m·n)   Space Complexity : O(min(m, n)/2)
 * ---------------------------------------------------------------------
 * DEEP DIVE — why the 4-pointer bound shrinking cannot skip cells
 *   The spiral partitions the grid into disjoint segments: every
 *   cell belongs to exactly one of the four sweeps of exactly one
 *   layer. Invariant: after processing layer k, all cells with
 *   k ≤ min(r, c, m-1-r, n-1-c) are visited. The layer index of any
 *   cell is min of its 4 edge distances — exactly what the guards
 *   track. Total work = Σ over layers of (perimeter) = m·n.
 * =====================================================================
 */

#include <iostream>       // std::cout — printing benchmark results
#include <string>         // std::string — formatting result lines
#include <vector>         // std::vector — matrix + spiral output
#include <unordered_map>  // (unused here; kept for uniform include set)
#include <unordered_set>  // (unused here; kept for uniform include set)
#include <chrono>         // std::chrono — steady_clock timing
#include <sys/resource.h> // getrusage — peak RAM measurement

/* =====================================================================
 * APPROACH 1 : Boundary shrinking with 4 pointers
 * ===================================================================== */
class Solution_1 {
public:
    /*
     * Purpose : Return all matrix elements in spiral order.
     * Inputs  : matrix — m×n vector<vector<int>>.
     * Output  : vector<int> with elements in spiral order.
     */
    std::vector<int> spiralOrder(const std::vector<std::vector<int>>& matrix) {
        std::vector<int> out;

        // The four shrinking boundaries of the current layer.
        int top = 0;
        int bottom = static_cast<int>(matrix.size()) - 1;
        int left = 0;
        int right = static_cast<int>(matrix[0].size()) - 1;

        // Keep peeling layers until the rectangle collapses.
        while (top <= bottom && left <= right) {

            // 1) top row, left → right.
            for (int c = left; c <= right; ++c) {
                out.push_back(matrix[top][c]);
            }
            ++top;   // top edge consumed → shrink from above

            // 2) right column, top → bottom.
            for (int r = top; r <= bottom; ++r) {
                out.push_back(matrix[r][right]);
            }
            --right; // right edge consumed → shrink from the right

            // 3) bottom row, right → left. Guarded: a single-row
            //    matrix (top > bottom now) must NOT double-visit.
            if (top <= bottom) {
                for (int c = right; c >= left; --c) {
                    out.push_back(matrix[bottom][c]);
                }
                --bottom;
            }

            // 4) left column, bottom → top. Guarded: a single-column
            //    matrix must NOT double-visit the remaining column.
            if (left <= right) {
                for (int r = bottom; r >= top; --r) {
                    out.push_back(matrix[r][left]);
                }
                ++left;
            }
        }
        return out;
    }
};

/* =====================================================================
 * APPROACH 2 : Layer-by-layer recursive peeling
 * ===================================================================== */
class Solution_2 {
public:
    /*
     * Purpose : Return all matrix elements in spiral order.
     * Inputs  : matrix — m×n vector<vector<int>>.
     * Output  : vector<int> with elements in spiral order.
     */
    std::vector<int> spiralOrder(const std::vector<std::vector<int>>& matrix) {
        std::vector<int> out;

        // Peel one layer, then recurse on the inner rectangle. The
        // recursion exits when the layer degenerates (empty rows/cols).
        peel(matrix, 0, static_cast<int>(matrix.size()) - 1,
             0, static_cast<int>(matrix[0].size()) - 1, out);
        return out;
    }

private:
    /*
     * Purpose : Append one spiral layer to `out`, then recurse inward.
     * Inputs  : matrix — the grid; t/b/l/r — current layer bounds;
     *           out — accumulator, filled in spiral order.
     * Output  : void (out is appended in place).
     */
    void peel(const std::vector<std::vector<int>>& matrix,
              int t, int b, int l, int r, std::vector<int>& out) const {
        if (t > b || l > r) {
            return;   // layer collapsed → whole spiral is emitted
        }

        for (int c = l; c <= r; ++c) { out.push_back(matrix[t][c]); }      // top →
        for (int row = t + 1; row <= b; ++row) { out.push_back(matrix[row][r]); } // right ↓
        if (t < b) { for (int c = r - 1; c >= l; --c) { out.push_back(matrix[b][c]); } } // bottom ←
        if (l < r) { for (int row = b - 1; row > t; --row) { out.push_back(matrix[row][l]); } } // left ↑

        // Recurse one layer deeper on every side.
        peel(matrix, t + 1, b - 1, l + 1, r - 1, out);
    }
};

/* =====================================================================
 * BENCHMARK — run both approaches on 3 matrices; time + peak RAM
 * ===================================================================== */
int main() {
    const std::vector<std::vector<int>> tests[] = {
        { { 1, 2, 3 }, { 4, 5, 6 }, { 7, 8, 9 } },
        { { 1, 2, 3, 4 }, { 5, 6, 7, 8 }, { 9, 10, 11, 12 } },
        { { 1 }, { 2 }, { 3 }, { 4 } },
    };
    const std::vector<int> expected[] = {
        { 1, 2, 3, 6, 9, 8, 7, 4, 5 },
        { 1, 2, 3, 4, 8, 12, 11, 10, 9, 5, 6, 7 },
        { 1, 2, 3, 4 },
    };

    bool all_ok = true;

    const auto run_all = [&](const std::string& label, const auto& solve) {
        struct rusage before, after;
        getrusage(RUSAGE_SELF, &before);
        const auto t0 = std::chrono::steady_clock::now();

        bool ok = true;
        for (size_t t = 0; t < 3; ++t) {
            const std::vector<int> got = solve(tests[t]);
            const bool pass = (got == expected[t]);
            ok = ok && pass;
            std::cout << "  test " << (t + 1) << " (" << tests[t].size() << "x"
                      << tests[t][0].size() << "): [";
            for (size_t i = 0; i < got.size(); ++i) {
                if (i) { std::cout << ","; }
                std::cout << got[i];
            }
            std::cout << "]" << (pass ? "  PASS" : "  FAIL") << "\n";
        }

        const auto t1 = std::chrono::steady_clock::now();
        getrusage(RUSAGE_SELF, &after);
        const double us = std::chrono::duration_cast<std::chrono::microseconds>(t1 - t0).count();
        const long kb = after.ru_maxrss - before.ru_maxrss;
        std::cout << label << " : " << (ok ? "PASS" : "FAIL")
                  << "  time = " << us << " us  peak-mem = " << kb << " KB\n";
        all_ok = all_ok && ok;
    };

    std::cout << "Test suite: 3 fixed matrices\n";
    run_all("Approach 1 (boundary shrinking)    ", [](const std::vector<std::vector<int>>& m) {
        return Solution_1().spiralOrder(m);
    });
    run_all("Approach 2 (recursive layer peel)  ", [](const std::vector<std::vector<int>>& m) {
        return Solution_2().spiralOrder(m);
    });

    std::cout << (all_ok
                  ? "PASS : both approaches produce the exact spiral order on all tests."
                  : "FAIL : at least one test failed.") << "\n";
    return all_ok ? 0 : 1;
}
